pkg/provider/provider.go

```diff
@@ -101,23 +112,57 @@ func DSN(s *schema.ResourceData) (string, error) {
 		region = ""
 	}
 
-	dsn, err := gosnowflake.DSN(&gosnowflake.Config{
-		Account:  account,
-		User:     username,
-		Region:   region,
-		Password: password,
-		Role:     role,
-	})
-
-	if browserAuth {
-		dsn, err = gosnowflake.DSN(&gosnowflake.Config{
-			Account:       account,
-			User:          username,
-			Region:        region,
-			Role:          role,
-			Authenticator: gosnowflake.AuthTypeExternalBrowser,
-		})
+	config := gosnowflake.Config{
+		Account: account,
+		User:    user,
+		Region:  region,
+		Role:    role,
 	}
 
-	return dsn, err
+	if len(privateKeyPath) != 0 {
+
+		rsaPrivateKey, err := ParsePrivateKey(privateKeyPath)
+		if err != nil {
+			return "", errors.Wrap(err, "Private Key could not be parsed")
+		}
+		config.PrivateKey = rsaPrivateKey
+		config.Authenticator = gosnowflake.AuthTypeJwt
+
+	} else if browserAuth {
+
+		config.Authenticator = gosnowflake.AuthTypeExternalBrowser
+
+	} else {
+		config.Password = password
+	}
+
+	return gosnowflake.DSN(&config)
+}
+
+func ParsePrivateKey(privateKeyPath string) (*rsa.PrivateKey, error) {
+	var err error
+
+	expandedPrivateKeyPath, err := homedir.Expand(privateKeyPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Invalid Path to private key")
+	}
+
+	privateKeyBytes, err := ioutil.ReadFile(expandedPrivateKeyPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not read private key")
+	}
+	if len(privateKeyBytes) == 0 {
+		return nil, errors.New("Private key is empty")
+	}
+
+	privateKey, err := ssh.ParseRawPrivateKey(privateKeyBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not parse private key")
+	}
+
+	rsaPrivateKey, ok := privateKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, errors.New("privateKey not of type RSA")
+	}
+	return rsaPrivateKey, err
```

Comment:
```suggestion
	return rsaPrivateKey, nil
```
at this point, err must be `nil` so lets make it explicit